test/src/test/java/hudson/model/QueueTest.java

```diff
@@ -855,6 +858,7 @@ public void shouldBeAbleToBlockFlyweightTaskAtTheLastMinute() throws Exception {
             throw new Exception("the upstream task could not be scheduled, thus the test will be interrupted");
         }
         //let s wait for the Upstream to enter the buildable Queue
+        Thread.sleep(1000);
         boolean enteredTheQueue = false;
         while (!enteredTheQueue) {
             for (Queue.BuildableItem item : Queue.getInstance().getBuildableItems()) {
```

Comment:
I think that deleting the section that waits for the build to enter the buildable queue ([Lines 862-870](https://github.com/jglick/jenkins/blob/eb5b40e25c720d2945aabfb4d0d776b491a84557/test/src/test/java/hudson/model/QueueTest.java#L862-L870)) would make the test consistent, as long as `scheduleBuild2`synchronously adds the build to the buildable queue before returning. If adding to the queue happens asynchronously, the test would still be flaky.

Reply:
Except that what we are trying to test is the dynamic and correct update of the queue, which unfortunately takes some time. 

1. Downstream enters the queue but it's waiting for executors
2. Upstream enters the queue
3. Upstream starts  #we need to wait for the upstream project to actually start so that we're sure the Queue has been updated and we can proceed to the newt step
4. Downstream leaves the queue cause gets blocked by upstream
5. Upstream completes
6. Downstream re-enters in the queue